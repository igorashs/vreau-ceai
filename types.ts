import FormidableLib from 'formidable';
import { ValidationError } from 'joi';

export type Product = {
  _id: string;
  name: string;
  category_id: string;
  src: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
};

export type ProductWithCategory = Product & {
  category_id: { name: string };
};

export type ErrorDetail = {
  message: string;
  name?: string;
};

export type ProductName = {
  name: string;
};

export type ProductNameErrorDetail = ErrorDetail & {
  name?: keyof ProductName;
};

export type ProductFields = ProductName & {
  src?: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
  category_id: string;
};

export type Category = {
  _id: string;
  name: string;
};

export type ProductErrorDetail = ErrorDetail & {
  name?: keyof ProductFields;
};

export type ApiResponse = {
  success: boolean;
  message: string;
  errors?: ErrorDetail[];
};

export type User = {
  _id: string;
  name: string;
  email: string;
  isManager: boolean;
};

export type UserEmail = {
  email: string;
};

export type UserEmailErrorDetail = ErrorDetail & {
  name?: keyof UserEmail;
};

export type UserLogin = UserEmail & { password: string };

export type UserSignup = UserEmail & {
  name: string;
  password: string;
  repeat_password: string;
};

export type UserPermissions = { isManager: boolean };

export type LabelMessage = { success: boolean; message: string };

export type UserAuth = {
  _id: string;
  name: string;
  isAdmin: boolean;
  isManager: boolean;
};

export type UserSession = {
  isAuth: boolean;
  user: UserAuth | null;
  needRefresh: boolean;
};

type dataType = 'data' | 'error' | 'field' | 'fileBegin' | 'file' | 'progress';

export declare class Formidable extends FormidableLib {
  emit(name: dataType, data: FormidableLib.EmitData | ValidationError): void;
}

export type CategoryName = {
  name: string;
};

export type CategoryNameErrorDetail = ErrorDetail & {
  name?: keyof CategoryName;
};

export type OrderSubmit = {
  tel: string;
  address: string;
};

export type OrderItem = {
  product_id: string;
  count: number;
};

export type OrderStatus = {
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
};

export type OrderStatusErrorDetail = ErrorDetail & {
  name?: keyof OrderStatus;
};

export type OrderNumber = {
  number: string;
};

export type OrderNumberErrorDetail = ErrorDetail & {
  name?: keyof OrderNumber;
};

export type OrderFields = {
  info: OrderSubmit;
  items: OrderItem[];
};

export type Order = {
  _id: string;
  user: {
    name: string;
    email: string;
  };
  number: string;
  total_price: number;
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
  items: {
    count: number;
    product: { name: string; price: number; quantity: number };
  }[];
  address: string;
  tel: string;
  orderedAt: Date;
  completedAt: Date;
};

export type UserOrders = { orders: Order[]; count: number };

export type ManagementOrders = { orders: Order[]; count: number };
