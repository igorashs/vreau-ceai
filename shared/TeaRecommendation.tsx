import styled from 'styled-components';
import Image from 'next/image';
import StyledLink from '@/shared/StyledLink';
import { ProductWithCategory } from 'types';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 1px solid var(--accent-dark);
  padding: calc(var(--baseline) / 2);
  background-color: #fff;
`;

const ImgContainer = styled.figure`
  width: 224px;
  height: 224px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid var(--accent-text-light);
`;

const Title = styled.h5`
  text-align: center;
`;

const Description = styled.p`
  align-self: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
`;

const Price = styled.p`
  text-align: center;
  grid-column: 1;
  font-weight: 500;
`;

const Actions = styled.div`
  display: grid;
`;

type TeaRecommendationProps = {
  tea: ProductWithCategory;
};

const TeaRecommendation = ({ tea }: TeaRecommendationProps) => {
  return (
    <Wrapper>
      <ImgContainer>
        <Image
          width={700}
          height={700}
          src={`/uploads/${tea.src}`}
          alt={tea.name}
        />
      </ImgContainer>
      <Title>{tea.name}</Title>
      <Description>{tea.description}</Description>
      <Price>{`${tea.price}lei - ${tea.quantity}g`}</Price>
      <Actions>
        <StyledLink
          href={`/categories/${tea.category_id.name || ''}/${tea.name}`}
          button
          label="vezi produs"
          text="vezi produs"
        />
      </Actions>
    </Wrapper>
  );
};

export default TeaRecommendation;
