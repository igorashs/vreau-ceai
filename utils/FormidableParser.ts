import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
