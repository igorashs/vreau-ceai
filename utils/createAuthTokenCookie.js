import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
