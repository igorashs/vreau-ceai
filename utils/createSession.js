import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
