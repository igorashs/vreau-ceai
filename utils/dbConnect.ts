import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
