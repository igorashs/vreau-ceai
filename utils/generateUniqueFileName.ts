import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
