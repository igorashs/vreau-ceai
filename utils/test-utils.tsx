import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
