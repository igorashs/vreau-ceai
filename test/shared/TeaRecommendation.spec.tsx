import { render } from '@/utils/test-utils';
import TeaRecommendation from '@/shared/TeaRecommendation';
import { ProductWithCategory } from 'types';
import { ImageProps } from 'next/image';

jest.mock('next/image', () => ({ alt, ...all }: ImageProps) => (
  <img {...all} alt={alt} />
));

describe('TeaRecommendation', () => {
  it('renders correctly', () => {
    const tea = {
      _id: '1',
      name: 'green-tea',
      description: 'tasty tea',
      recommend: true,
      src: 'tea.png',
      price: 39,
      quantity: 50,
      total_quantity: 500,
      category_id: {
        name: 'green',
      },
    } as ProductWithCategory;

    const { container, getByAltText, getByText, getByRole } = render(
      <TeaRecommendation tea={tea} />,
    );

    expect(getByAltText(tea.name)).toHaveAttribute(
      'src',
      expect.stringContaining(tea.src),
    );

    expect(getByText(tea.name)).toBeVisible();
    expect(getByText(tea.description)).toBeVisible();
    expect(getByText(`${tea.price}lei`, { exact: false })).toBeVisible();
    expect(getByText(`${tea.quantity}g`, { exact: false })).toBeVisible();

    expect(getByRole('link')).toHaveAttribute(
      'href',
      expect.stringContaining(`${tea.category_id.name}/${tea.name}`),
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
