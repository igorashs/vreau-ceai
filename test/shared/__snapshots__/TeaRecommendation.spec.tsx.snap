// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeaRecommendation renders correctly 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c6:hover {
  background-color: var(--accent);
}

.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 1px solid var(--accent-dark);
  padding: calc(var(--baseline) / 2);
  background-color: #fff;
}

.c1 {
  width: 224px;
  height: 224px;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid var(--accent-text-light);
}

.c2 {
  text-align: center;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
}

.c4 {
  text-align: center;
  grid-column: 1;
  font-weight: 500;
}

.c5 {
  display: grid;
}

<div
  class="c0"
>
  <figure
    class="c1"
  >
    <img
      alt="green-tea"
      height="700"
      src="tea.png"
      width="700"
    />
  </figure>
  <h5
    class="c2"
  >
    green-tea
  </h5>
  <p
    class="c3"
  >
    tasty tea
  </p>
  <p
    class="c4"
  >
    39lei - 50g
  </p>
  <div
    class="c5"
  >
    <a
      aria-label="vezi produs"
      class="c6"
      href="/categories/green/green-tea"
      rel=""
      target=""
    >
      vezi produs
    </a>
  </div>
</div>
`;
